internal/scheduler/schedulercontexts.go

```diff
@@ -0,0 +1,364 @@
+package scheduler
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"text/tabwriter"
+	"time"
+
+	"golang.org/x/exp/maps"
+	v1 "k8s.io/api/core/v1"
+
+	armadamaps "github.com/armadaproject/armada/internal/common/maps"
+	armadaslices "github.com/armadaproject/armada/internal/common/slices"
+	"github.com/armadaproject/armada/internal/scheduler/schedulerobjects"
+)
+
+// SchedulingContext captures the decisions made by the scheduler during one invocation.
+type SchedulingContext struct {
+	// Time at which the scheduling cycle started.
+	Started time.Time
+	// Time at which the scheduling cycle finished.
+	Finished time.Time
+	// ExecutorId for which the scheduler was invoked.
+	ExecutorId string
+	// Per-queue scheduling contexts.
+	QueueSchedulingContexts map[string]*QueueSchedulingContext
+	// Total resources across all clusters available at the start of the scheduling cycle.
+	TotalResources schedulerobjects.ResourceList
+	// Resources assigned across all queues during this scheduling cycle.
+	ScheduledResourcesByPriority schedulerobjects.QuantityByPriorityAndResourceType
+	// Total number of jobs successfully scheduled in this round.
+	NumScheduledJobs int
+	// Reason for why the scheduling round finished.
+	TerminationReason string
+	// Protects everything in this struct.
```

Comment:
Why does this need to be threadsafe?  I would have assumed that this is a simple record type that we pass round the single-threaded scheduling cycle.

Reply:
Also all the fields in this struct are exported but they need to be protected by a mutex.  I don't think that will be obvious to callers.

Reply:
(above 2 comments apply to the other contexts below)

Reply:
Those mutexes are left over from when I was planning out how to parallelise the scheduler (which I never ended up doing). I've removed these mutexes.